accesscontrol/jwt.go

```diff
@@ -322,81 +319,106 @@ func (j *JWT) validateClaims(token *jwt.Token, claims map[string]interface{}) (m
 	return tokenClaims, nil
 }
 
-const errValueMsg = "value of %s claim must either be a string containing a space-separated list of %s or a list of string %s"
+func (j *JWT) getScopeValues(tokenClaims map[string]interface{}, log *logrus.Entry) []string {
+	var scopeValues []string
 
-var errScopeValue = fmt.Errorf(errValueMsg, "scope", "scope values", "scope values")
-var errRolesValue = fmt.Errorf(errValueMsg, "roles", "roles", "roles")
+	scopeValues = j.addScopeValueFromScope(tokenClaims, scopeValues, log)
 
-func (j *JWT) getScopeValues(tokenClaims map[string]interface{}) ([]string, error) {
-	var scopeValues []string
+	scopeValues = j.addScopeValueFromRoles(tokenClaims, scopeValues, log)
 
-	if j.scopeClaim != "" {
-		scopesFromClaim, exists := tokenClaims[j.scopeClaim]
-		if !exists {
-			return nil, fmt.Errorf("missing expected scope claim %q", j.scopeClaim)
-		}
+	return scopeValues
+}
 
-		// ["foo", "bar"] is stored as []interface{}, not []string, unfortunately
-		scopesArray, ok := scopesFromClaim.([]interface{})
-		if ok {
-			for _, v := range scopesArray {
-				s, ok := v.(string)
-				if !ok {
-					return nil, errScopeValue
-				}
-				scopeValues = addScopeValue(scopeValues, s)
-			}
-		} else {
-			scopesString, ok := scopesFromClaim.(string)
+const warnInvalidValueMsg = "invalid %s claim value type, ignoring claim, value: %#v"
+
+func (j *JWT) addScopeValueFromScope(tokenClaims map[string]interface{}, scopeValues []string, log *logrus.Entry) []string {
+	if j.scopeClaim == "" {
+		return scopeValues
+	}
+
+	scopesFromClaim, exists := tokenClaims[j.scopeClaim]
+	if !exists {
+		return scopeValues
+	}
+
+	// ["foo", "bar"] is stored as []interface{}, not []string, unfortunately
+	scopesArray, ok := scopesFromClaim.([]interface{})
+	if ok {
+		var vals []string
+		for _, v := range scopesArray {
+			s, ok := v.(string)
 			if !ok {
-				return nil, errScopeValue
-			}
-			for _, s := range strings.Split(scopesString, " ") {
-				scopeValues = addScopeValue(scopeValues, s)
+				log.Warn(fmt.Sprintf(warnInvalidValueMsg, "scope", scopesFromClaim))
+				return scopeValues
```

Comment:
```suggestion
				continue
```

Reply:
No, that would cause
```
		for _, val := range vals {
			scopeValues = addScopeValue(scopeValues, val)
		}
```
to be executed, so that all the previously found string values (which are currently in `vals`) are used. But for mixed arrays the _complete_ value is to be rejected. Same with roles claim.